File: apps/api-flask/src/services/audit_service.py
"""Audit logging service"""
from ..app import db
from ..models.audit_log import AuditLog


class AuditService:
    """Service for audit logging"""

    @staticmethod
    def log_auth(action: str, user_id: str = None, metadata: dict = None,
                 ip_address: str = None, user_agent: str = None):
        """Log authentication event"""
        log = AuditLog(
            user_id=user_id,
            action=action,
            entity_type='auth',
            meta=metadata or {},
            ip_address=ip_address,
            user_agent=user_agent
        )
        db.session.add(log)
        db.session.commit()

    @staticmethod
    def log_action(action: str, user_id: str, entity_type: str = None,
                   entity_id: str = None, metadata: dict = None,
                   ip_address: str = None, user_agent: str = None):
        """Log user action"""
        log = AuditLog(
            user_id=user_id,
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            meta=metadata or {},
            ip_address=ip_address,
            user_agent=user_agent
        )
        db.session.add(log)
        db.session.commit()

File: apps/api-spring/src/main/java/com/todoapp/application/service/AuditService.java
package com.todoapp.application.service;

import com.todoapp.domain.repository.AuditLogRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuditService {

    private final AuditLogRepository auditLogRepository;

    @Async
    @Transactional
    public void logAction(UUID userId, String action, String metadata, 
                         String ipAddress, String userAgent) {
        try {
            auditLogRepository.createLog(userId, action, metadata, ipAddress, userAgent);
            log.debug("Audit log created: user={}, action={}", userId, action);
        } catch (Exception e) {
            log.error("Failed to create audit log", e);
            // Don't throw - audit failures shouldn't break business operations
        }
    }

    public List<AuditLogRepository.AuditLog> getUserAuditLogs(UUID userId, int limit, int offset) {
        return auditLogRepository.findByUserId(userId, limit, offset);
    }

    public List<AuditLogRepository.AuditLog> getAllAuditLogs(int limit, int offset) {
        return auditLogRepository.findAll(limit, offset);
    }
}
